aten/src/ATen/native/sparse/cuda/SparseMatMul.cu

```diff
@@ -0,0 +1,608 @@
+  auto n = mat2.size(1);
+
+  TORCH_CHECK((m <= INT_MAX) && (n <= INT_MAX) && (k1 <= INT_MAX),
+    "At the moment, cusparseDcsrgemm2 only supports m, n, k, nnz with the bound [val] <= ", INT_MAX, ".",
+    "If you need this, please file an issue on GitHub."
+  );
+
+  csrMatrixRef<scalar_t> csr_mat1(
+      mat1_indices.data_ptr<int>(),
+      mat1_indptr.data_ptr<int>(),
+      mat1_values.data_ptr<scalar_t>(),
+      (int)mat1._nnz(),
+      {(int)mat1.size(0), (int)mat1.size(1)});
+
+  csrMatrixRef<scalar_t> csr_mat2(
+      mat2_indices.data_ptr<int>(),
+      mat2_indptr.data_ptr<int>(),
+      mat2_values.data_ptr<scalar_t>(),
+      (int)mat2._nnz(),
+      {(int)mat2.size(0), (int)mat2.size(1)});
+
+  auto output_indices = result._indices();
+  auto output_values = result._values();
+
+  // Sparse matrix multiplication
+  csrOutput csr_output = cuSparse_matrix_multiply<scalar_t>(csr_mat1, csr_mat2, output_values, output_indices); 
+  auto nnz = csr_output.nnz_;
+
+  output_values.set_(csr_output.csr_values_);
+  output_indices.resize_({2, nnz});
+  auto output_indices_accessor = output_indices.packed_accessor<int64_t, 2>();
+
+  auto csr_output_pointers_accessor =
+      csr_output.csr_pointers_.packed_accessor<int, 1>();
+
+  auto csr_output_ind_accessor =
+      csr_output.csr_indices_.packed_accessor<int, 1>();
+
+  auto major_dim = result.size(0);
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+
+  thrust::for_each(
+      policy,
+      thrust::make_counting_iterator(int64_t(0)),
+      thrust::make_counting_iterator(int64_t(major_dim)),
+      [output_indices_accessor,
+       csr_output_pointers_accessor,
+       major_dim,
+       nnz] __device__(int64_t i) {
+        auto Ap = csr_output_pointers_accessor.data();
+        int64_t* indices_row = output_indices_accessor[0].data();
+
+        for (int jj = Ap[i];  jj < Ap[i + 1]; jj++) {
+          indices_row[jj] = i;
+        }
+      });
+
+  thrust::for_each(
+    policy,
+    thrust::make_counting_iterator(int64_t(0)),
+    thrust::make_counting_iterator(int64_t(csr_output.nnz_)),
+    [output_indices_accessor,
+      csr_output_pointers_accessor,
+      csr_output_ind_accessor,
+      major_dim,
+      nnz] __device__(int64_t i) {
+      int64_t* indices_col = output_indices_accessor[1].data();
+      indices_col[i] = csr_output_ind_accessor[i];
+    });
+}
+
+template <typename scalar_t, bool grad_by_row>
+void sparse_matmul_kernel_grad(Tensor& output, const Tensor& grad, const Tensor& x) {
+  /* 
+    Computes  the backward output  for matrix C = A*B.
+
+    C = A@B 
+      then 
+    A_grad = C_grad @ B^T
+    B_grad = A^T @ C_grad
+
+    if grad_by_row == true:
+      output = x^T @ C_grad 
+    else:
+      output = C_grad @ x^T 
+  */
+  Tensor grad_filled = at::ones(grad.sizes(), grad.options().layout(kStrided));
```

Comment:
I don't think that's correct to use `at::ones` for derivative calculations. `grad` should be used instead of `grad_filled`.